File: old/commands/cleanup.py
"""Cleanup command orchestration."""
from __future__ import annotations
import sys
from dataclasses import dataclass
from typing import List
from cli import PCT, Find
from libs import common
from libs.logger import get_logger
from libs.command import Command
from services.lxc import LXCService
from services.pct import PCTService
logger = get_logger(__name__)
destroy_container = common.destroy_container

class CleanupError(RuntimeError):
    """Raised when cleanup fails."""
@dataclass

class Cleanup(Command):
    """Holds cleanup context."""
    lxc_service: LXCService = None
    pct_service: PCTService = None

    def run(self, args):
        """Execute the cleanup workflow."""
        import traceback
        try:
            logger.info("=" * 50)
            logger.info("Cleaning Up Lab Environment")
            logger.info("=" * 50)

            logger.info("Destroying ALL containers and templates...")

            # Connect LXC service (injected via DI)
            if not self.lxc_service.connect():
                logger.error("Failed to connect to Proxmox host %s", self.cfg.proxmox_host)
                raise CleanupError("Failed to connect to Proxmox host")

            try:
                self._destroy_containers()
                self._remove_templates()
            finally:
                if self.lxc_service:
                    self.lxc_service.disconnect()
        except CleanupError as err:
            logger.error("Error during cleanup: %s", err)
            logger.error(traceback.format_exc())
            sys.exit(1)

    def _destroy_containers(self):
        logger.info("Stopping and destroying containers...")

        container_ids = self._list_container_ids()
        total = len(container_ids)

        if total == 0:
            logger.info("No containers found")
            return

        logger.info("Found %s containers to destroy: %s", total, ", ".join(container_ids))

        for idx, cid in enumerate(container_ids, 1):
            logger.info("[%s/%s] Processing container %s...", idx, total, cid)
            destroy_container(self.cfg.proxmox_host, cid, cfg=self.cfg, lxc_service=self.lxc_service)

        self._verify_containers_removed()

    def _list_container_ids(self) -> List[str]:
        list_cmd = PCT().status()
        result, exit_code = self.lxc_service.execute(list_cmd)
        container_ids: List[str] = []

        if result:
            lines = result.strip().split("\n")

            for line in lines[1:]:
                parts = line.split()

                if parts and parts[0].isdigit():
                    container_ids.append(parts[0])

        return container_ids

    def _verify_containers_removed(self):
        logger.info("Verifying all containers are destroyed...")

        remaining_result, exit_code = self.lxc_service.execute(PCT().status())
        remaining_ids: List[str] = []

        if remaining_result:
            remaining_lines = remaining_result.strip().split("\n")
            for line in remaining_lines[1:]:
                parts = line.split()
                if parts and parts[0].isdigit():
                    remaining_ids.append(parts[0])

        if remaining_ids:
            raise CleanupError(f"{len(remaining_ids)} containers still exist: {', '.join(remaining_ids)}")

        logger.info("All containers destroyed")

    def _remove_templates(self):
        logger.info("Removing templates...")
        template_dir = self.cfg.proxmox_template_dir

        logger.info("Cleaning template directory %s...", template_dir)
        count_cmd = Find().directory(template_dir).maxdepth(1).type("f").name("*.tar.zst").count()
        count_result, exit_code = self.lxc_service.execute(count_cmd)
        template_count = count_result.strip() if count_result else "0"

        logger.info("Removing %s template files...", template_count)
        delete_cmd = Find().directory(template_dir).maxdepth(1).type("f").name("*.tar.zst").delete()
        self.lxc_service.execute(delete_cmd)

        logger.info("Templates removed")

File: src/commands/cleanup.go
package commands

import (
	"enva/cli"
	"enva/libs"
	"enva/services"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// CleanupError is raised when cleanup fails
type CleanupError struct {
	Message string
}

func (e *CleanupError) Error() string {
	return e.Message
}

// Cleanup handles cleanup operations
type Cleanup struct {
	cfg        *libs.LabConfig
	lxcService *services.LXCService
	pctService *services.PCTService
}

// NewCleanup creates a new Cleanup command
func NewCleanup(cfg *libs.LabConfig, lxcService *services.LXCService, pctService *services.PCTService) *Cleanup {
	return &Cleanup{
		cfg:        cfg,
		lxcService: lxcService,
		pctService: pctService,
	}
}

// Run executes the cleanup
func (c *Cleanup) Run() error {
	logger := libs.GetLogger("cleanup")
	defer func() {
		if r := recover(); r != nil {
			if err, ok := r.(error); ok {
				logger.Error("Error during cleanup: %s", err.Error())
				logger.LogTraceback(err)
			} else {
				logger.Error("Error during cleanup: %v", r)
			}
			os.Exit(1)
		}
	}()
	logger.Info("==================================================")
	logger.Info("Cleaning Up Lab Environment")
	logger.Info("==================================================")
	logger.Info("Destroying ALL containers and templates...")
	if !c.lxcService.Connect() {
		logger.Error("Failed to connect to LXC host %s", c.cfg.LXCHost())
		return &CleanupError{Message: "Failed to connect to LXC host"}
	}
	defer c.lxcService.Disconnect()
	if err := c.destroyContainers(); err != nil {
		return err
	}
	if err := c.removeTemplates(); err != nil {
		return err
	}
	return nil
}

func (c *Cleanup) destroyContainers() error {
	logger := libs.GetLogger("cleanup")
	logger.Info("Stopping and destroying containers...")
	containerIDs := c.listContainerIDs()
	total := len(containerIDs)
	if total == 0 {
		logger.Info("No containers found")
		return nil
	}
	logger.Info("Found %d containers to destroy: %s", total, strings.Join(containerIDs, ", "))
	for idx, cidStr := range containerIDs {
		logger.Info("[%d/%d] Processing container %s...", idx+1, total, cidStr)
		cid, err := strconv.Atoi(cidStr)
		if err != nil {
			logger.Info("Invalid container ID: %s", cidStr)
			continue
		}
		libs.DestroyContainer(c.cfg.LXCHost(), cid, c.cfg, c.lxcService)
	}
	return c.verifyContainersRemoved()
}

func (c *Cleanup) listContainerIDs() []string {
	listCmd := cli.NewPCT().Status()
	result, _ := c.lxcService.Execute(listCmd, nil)
	var containerIDs []string
	if result != "" {
		lines := strings.Split(strings.TrimSpace(result), "\n")
		for _, line := range lines[1:] {
			parts := strings.Fields(line)
			if len(parts) > 0 && isNumeric(parts[0]) {
				containerIDs = append(containerIDs, parts[0])
			}
		}
	}
	return containerIDs
}

func (c *Cleanup) verifyContainersRemoved() error {
	logger := libs.GetLogger("cleanup")
	logger.Info("Verifying all containers are destroyed...")
	remainingResult, _ := c.lxcService.Execute(cli.NewPCT().Status(), nil)
	var remainingIDs []string
	if remainingResult != "" {
		remainingLines := strings.Split(strings.TrimSpace(remainingResult), "\n")
		for _, line := range remainingLines[1:] {
			parts := strings.Fields(line)
			if len(parts) > 0 && isNumeric(parts[0]) {
				remainingIDs = append(remainingIDs, parts[0])
			}
		}
	}
	if len(remainingIDs) > 0 {
		return &CleanupError{Message: fmt.Sprintf("%d containers still exist: %s", len(remainingIDs), strings.Join(remainingIDs, ", "))}
	}
	logger.Info("All containers destroyed")
	return nil
}

func (c *Cleanup) removeTemplates() error {
	logger := libs.GetLogger("cleanup")
	logger.Info("Removing templates...")
	templateDir := c.cfg.LXCTemplateDir()
	logger.Info("Cleaning template directory %s...", templateDir)
	countCmd := cli.NewFind().Directory(templateDir).Maxdepth(1).Type("f").Name("*.tar.zst").Count()
	countResult, _ := c.lxcService.Execute(countCmd, nil)
	templateCount := "0"
	if countResult != "" {
		templateCount = strings.TrimSpace(countResult)
	}
	logger.Info("Removing %s template files...", templateCount)
	deleteCmd := cli.NewFind().Directory(templateDir).Maxdepth(1).Type("f").Name("*.tar.zst").Delete()
	c.lxcService.Execute(deleteCmd, nil)
	logger.Info("Templates removed")
	return nil
}

func isNumeric(s string) bool {
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return len(s) > 0
}
